plugin/evm/mempool.go

```diff
@@ -0,0 +1,210 @@
+// (c) 2019-2020, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package evm
+
+import (
+	"sync"
+
+	"github.com/ava-labs/avalanchego/cache"
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+// Mempool is a simple mempool for atomic transactions
+type Mempool struct {
+	lock sync.Mutex
+	// maxSize is the maximum number of transactions allowed to be kept in mempool
+	maxSize int
+	// currentTx is the transaction about to be added to a block.
+	currentTx *Tx
+	// txs is the set of transactions that need to be issued into new blocks
+	txs map[ids.ID]*Tx
+	// issuedTxs is the set of transactions that have been issued into a new block
+	issuedTxs map[ids.ID]*Tx
+	// discardedTxs is an LRU Cache of recently discarded transactions
+	discardedTxs *cache.LRU
+	// Pending is a channel that signals when the mempool is ready to put an atomic transaction
+	// into a new block
+	Pending chan struct{}
+}
+
+// NewMempool returns a Mempool with [maxSize]
+func NewMempool(maxSize int) *Mempool {
+	return &Mempool{
+		txs:          make(map[ids.ID]*Tx),
+		issuedTxs:    make(map[ids.ID]*Tx),
+		discardedTxs: &cache.LRU{Size: 10},
+		Pending:      make(chan struct{}, 1),
+		maxSize:      maxSize,
+	}
+}
+
+// Len returns the number of transactions in the mempool
+func (m *Mempool) Len() int {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return m.length()
+}
+
+// assumes the lock is held
+func (m *Mempool) length() int {
+	return len(m.txs) + len(m.issuedTxs)
+}
+
+// Add attempts to add [tx] to the mempool and returns true if the
+// mempool requires a new block to be built.
+func (m *Mempool) AddTx(tx *Tx) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	txID := tx.ID()
+	// If [txID] has already been issued, there is no need to
+	// build a new block.
+	if _, exists := m.issuedTxs[txID]; exists {
+		return nil
+	}
+
+	// Each of the following cases results in there being a
+	// pending transaction, so we add to the channel here.
+	m.addPending()
```

Comment:
Would pending not already have been sent in the following places where we return early? Why send it again?

Reply:
Defense in depth, but you're right there should already be an item in each of the cases.